### PR TITLE
feat(model): build HFDC constraints once per parameter at model level

### PR DESCRIPTION
**Pull request by kratsg**
## Summary

- **Bug fixed**: When multiple `HistFactoryDistChannel` (HFDC) instances share nuisance parameters (e.g. `lumi` correlated across channels), each channel's `extended_likelihood` previously emitted its own constraint term, causing double-counting in the joint NLL.
- `Model._build_distribution_node` now stores the Poisson-only term in `self._hfdc_poisson` and collects deduplicated constraint expressions in `self._hfdc_constraints`. The full expression (Poisson x constraints) is still stored in `self.distributions` so `model.logpdf()` is unchanged.
- `Model.log_prob` uses `self._hfdc_poisson[dist_name]` for binned channels and appends the deduplicated constraint log-terms in a single pass after the main loop.
- `Model._build_parameter_node` bakes `{name}_observed` as `pt.constant` when a paired `BinnedData` exists in the likelihood, removing it from the free-input set for JAX transpilation.
- `Workspace._validate_hfdc_constraints` validates constraint-type consistency across channels and forbids `shapesys`/`staterror` parameters from being shared (they are per-channel by design).

## Relationship to other PRs

Depends on #220 (histosys nominal-rate bug fix) — the real-world NLL comparisons in the slow test suite are only meaningful once the histosys bug is resolved.

## Test plan

- [ ] `pixi run --environment py312 test tests/test_hfdc_model_constraints.py` -- 15 tests, all green
- [ ] `pixi run --environment py312 test` -- 926 passed, no regressions
- [ ] `pixi run pre-commit` -- all hooks pass
- [ ] `pixi run --environment py312 test-slow tests/test_realworld.py -s` -- multi-channel workspaces (requires #220 merged first)

Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added constraint deduplication for multi-channel analyses, ensuring shared constraints are applied once rather than multiple times across channels.
  * Introduced workspace validation to detect and prevent inconsistent constraint configurations across channels.

* **Bug Fixes**
  * Corrected constraint accumulation in likelihood calculations to avoid duplicate constraint terms.

* **Tests**
  * Added comprehensive test coverage for constraint deduplication and validation logic, including edge cases with shared and independent constraints.

[Review Change Stack](https://app.coderabbit.ai/change-stack/scipp-atlas/pyhs3/pull/222)

### CODE REVIEW

File: src/pyhs3/workspace.py [hunk 1]
```diff
@@ -127,6 +139,64 @@ def _resolve_fk_list(
...
+                        for param in modifier.parameters:
+                            if param in owners and owners[param] != dist_obj.name:
+                                kind = modifier.type
+                                msg = (
+                                    f"{kind} parameter '{param}' appears in both "
+                                    f"'{owners[param]}' and '{dist_obj.name}'; "
+                                    f"{kind} is per-channel and may not be shared."
+                                )
+                                raise ValueError(msg)
+                            owners[param] = dist_obj.name
```
[the remark this answers is not in the document]
Comment (kratsg) [reply src/pyhs3/workspace.py]: not needed

File: src/pyhs3/model.py [hunk 2]
```diff
@@ -256,6 +273,11 @@ def log_prob(self) -> TensorVar:
                 if aux_name in self.distributions:
                     terms.append(pt.log(self.distributions[aux_name]))
 
+        # HFDC constraint terms: collected once per unique nuisance parameter
+        # across all channels during graph construction.
+        for constraint_expr in self._hfdc_constraints:
+            terms.append(pt.log(constraint_expr))
```
[the remark this answers is not in the document]
Comment (kratsg) [reply src/pyhs3/model.py]: addressed

File: tests/test_model.py [hunk 3]
```diff
@@ -738,7 +738,7 @@ def test_workspace_loads_likelihoods_correctly(
         assert len(combined_likelihood.data) == 2
         assert combined_likelihood.data[0].name == "observed_data"
         assert combined_likelihood.data[1].name == "observed_data"
-        assert combined_likelihood.aux_distributions == ["background_dist"]
+        assert combined_likelihood.aux_distributions[0].name == "background_dist"
```
[the remark this answers is not in the document]
Comment (kratsg) [reply tests/test_model.py:742]: Fixed. Added assert len(combined_likelihood.aux_distributions) == 1 before the name check to catch spurious extra entries.

File: src/pyhs3/workspace.py [hunk 4]
```diff
@@ -127,6 +139,79 @@ def _resolve_fk_list(
...
+            for param in multi_mod.parameters:
+                if param in owners and owners[param] != channel_name:
+                    kind = multi_mod.type
+                    msg = (
+                        f"{kind} parameter '{param}' appears in both "
+                        f"'{owners[param]}' and '{channel_name}'; "
+                        f"{kind} is per-channel and may not be shared."
+                    )
+                    raise ValueError(msg)
+                owners[param] = channel_name
```
[the remark this answers is not in the document]
Comment (kratsg) [reply src/pyhs3/workspace.py:180]: The early-return guard modifier.constraint is None is unreachable: HasConstraint.constraint is typed as a non-Optional Literal["Gauss", "Poisson", "LogNormal"], so any object that passes isinstance(modifier, HasConstraint) will always have constraint set.

File: src/pyhs3/model.py [hunk 5]
```diff
@@ -280,8 +322,34 @@ def _ensure_array(
...
+                return pt.constant(
+                    np.asarray(datum.contents, dtype=np.float64), name=node_name
+                )
+        return None
+
     def _build_parameter_node(self, node_name: str, context: Context) -> TensorVar:
         """Build a parameter node: baked constant (const=True) or bounded free variable."""
+        baked = self._try_bake_hfdc_observed(node_name)
+        if baked is not None:
+            return baked
```
[the remark this answers is not in the document]
Comment (kratsg) [reply src/pyhs3/model.py:351]: Skipped. free_params iterates self.parameterset only; {channel}_observed parameters are baked as pt.constant in self.parameters, never in the ParameterSet, so they can't appear in free_params.

File: src/pyhs3/model.py [hunk 6]
```diff
@@ -252,9 +283,20 @@ def log_prob(self) -> TensorVar:
...
+                    ]
+                )
+                for aux_name in self._likelihood.aux_distributions
+            )
+
+        # HFDC constraint terms: collected once per unique nuisance parameter
+        # across all channels during graph construction.
+        terms.extend(
+            pt.log(constraint_expr) for constraint_expr in self._hfdc_constraints
+        )
```
[the remark this answers is not in the document]
Comment (kratsg) [reply src/pyhs3/model.py:299]: Fixed. _build_distribution_node now gates constraint collection behind a membership check against self._likelihood.distributions. The bug: constraints from HFDC channels not in the active likelihood were unconditionally added to log_prob. Test confirmed RED before the fix and GREEN after.
